src/promql/src/extension_plan/series_divide.rs

```diff
@@ -255,21 +255,20 @@ impl Stream for SeriesDivideStream {
 
     fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
         loop {
-            // It has to be cloned here, otherwise the later ready! will mess things up
-            if let Some(batch) = self.buffer.clone() {
+            if let Some(batch) = self.buffer.as_ref() {
                 let same_length = self.find_first_diff_row(&batch) + 1;
                 if same_length >= batch.num_rows() {
                     let next_batch = match ready!(self.as_mut().fetch_next_batch(cx)) {
-                        Some(Ok(batch)) => batch,
+                        Some(Ok(next_batch)) => next_batch,
                         None => {
-                            self.buffer = None;
                             self.num_series += 1;
+                            let batch = self.buffer.take().expect("must be some batch");
                             return Poll::Ready(Some(Ok(batch)));
                         }
                         error => return Poll::Ready(error),
                     };
-                    let new_batch =
-                        compute::concat_batches(&batch.schema(), &[batch.clone(), next_batch])?;
+                    let batch = self.buffer.take().expect("must be some batch");
```

Comment:
I don't like this `expect`. But seems not better style ...

We don't mutate buffer in `fetch_next_batch` and in an `if Some` branch, so the buffer must be present.

Reply:
Or we can somehow "handle" the None case.

Reply:
Add a new commit to handle it. You can compare.